// webtap/index.js
var http = require('http');
var tap = require('tap');

module.exports = webtap;

function webtap (server) {
  if (typeof server == 'function') server = http.createServer(server);

  return function test (name, method, path, cb) {
    var args = [].slice.call(arguments);
    var cb = args.pop();
    var path = args.pop();
    var method = args.pop();
    var name = args.pop();

    tap.test(name || method + ' ' + path, function (t) {
      server.listen(function () {
        var opts = {
          hostname : 'localhost',
          port : server.address().port,
          path : path,
          method : method
        }
        var req = http.request(opts, function (res) {
          t.body = body;
          cb(t, res);

          function body (str, msg) {
            var data = '';
            res.on('data', function (d) { data += d });
            res.on('end', function () {
              t.equal(data, str, msg);
              server.close(t.end.bind(t));
            });
          }
        })
        req.on('error', t.notOk.bind(t));
        req.end();
      });
    });
  }
}
